text/0000-project-descriptor.md

```diff
@@ -0,0 +1,353 @@
+
+# Motivation
+[motivation]: #motivation
+
+This proposal is meant to address all of the following issues:
+
+* Build metadata (i.e. information about the code being built)
+* [Monorepos](https://github.com/buildpack/spec/issues/39)
+* [Application descriptor](https://github.com/buildpack/spec/issues/44)
+* Inline buildpacks
+* [Ignoring files](https://github.com/buildpack/pack/issues/210)
+* Multiple images (i.e. one build, multiple images as output)
+* Custom launch configuration
+* Codified buildpacks
+
+# What it is
+[what-it-is]: #what-it-is
+
+Terminology:
+
+* **project**: a repository containing source code for an app, service, function, buildpack or a monorepo containing any combination of those.
+* **image**: the output of a running a buildpack(s) against a project
+* **inline buildpack**: a buildpack that is part of a project's source code
+
+The target personas for this proposal is buildpack users who need to enrich or configure buildpack execution. The new elements in  `buildpack.toml` support many different use cases including two new top level tables:
+
+- `[project]`: (optional) defines configuration for a project
+- `[[images]]`: (optional) defines configuration for an image
+- `[metadata]`: (optional) metadata about the repository
+
+## `[project]`
+
+The top-level `[project]` table may contain configuration about the repository, including `id` and `version`, but also metadata about how it is authored, documented, and version controlled. If any of these values are redefined in `[buildpack]` or `[[images]]` they will override the values in `[project]`.
+
+The `project.id`
+
+```toml
+[project]
+id = "<string>"
+name = "<string>"
+version = "<string>"
+authors = ["<string>"]
+documentation = "<string>"
+license = "<string>"
+source = "<string>"
+```
+
+## `[project.include]` and `[project.exclude]`
+
+A list of files to include in the build (while excluding everything else):
+
+```toml
+[project]
+include = [
+    "cmd/",
+    "go.mod",
+    "go.sum",
+    "*.go"
+]
+```
+
+A list of files to exclude from the build (while including everything else)
+
+```toml
+[project]
+exclude = [
+    "spec/"
+]
+```
+
+The `.gitignore` pattern is used in both cases. The `exclude` and `include` keys are mutually exclusive, and if both are present the Lifecycle will error out. These lists apply to both buildpacks built with `pack create-package` and apps built with `pack build`.
+
+## `[project.publish]`
+
+A boolean that, when set to `false`, will prevent a platform from publishing the image generated from a build to a registry.
+
+## `[[project.buildpacks]]`
+
+The `[[project.buildpacks]]` array defines the buildpacks that will run against the app:
+
+```toml
+[[project.buildpacks]]
+id = "<buildpack ID (required)>"
+version = "<buildpack version (optional default=latest)>"
+optional = "<bool (optional default=false)>"
+uri = "<url or path to the buildpack (optional default=urn:buildpack:<id>)"
```

Comment:
Should this be an image ref to a buildpackage? Otherwise, I think we'd need a centralized buildpack registry to resolve all the dependent IDs in the downloaded artifact.